text/unification.md

```diff
@@ -0,0 +1,268 @@
+
+A unification algorithm turns a `Σ` and a collection of unification problems into either an impossibility of satisfaction of the equations, or in a collection of (well-typed) definitions for the existential variables such that all problems are satisfied, or a residual set of unification problems that the algorithm is neither able to invalidate nor to satisfy (since unification in general is undecidable in the Calculus of Inductive Constructions).
+
+## Reflexive unification problems
+
+We call *reflexive unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= ?e@{subst'}`.
+
+## Atomic unification problems
+
+We call *atomic unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= ?e'@{subst'}` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? ?e'@{subst'}`.
+
+## Simple unification problems
+
+We call *simple unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= t` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? t`. Unification basically reduces arbitrary complex unification problems to simple ones which in turn may generate new arbitrary complex problems.
+
+## Simplification of reflexive unification problems
+
+A reflexive unification problem induces constraints on the possible definitions of `?e`: once `?e` is defined, it may generate new unification problems between terms of the instances that contain existential variables.
+
+A reflexive unification problem can be simplified as follows:
+- Variables of the context whose instances in the substitution are provably incompatible (e.g. because corresponding to distinct head normal forms) can be removed from the context.
+- If both instances are free of existential variables, the problem can be dropped since its resolution will eventually only create equations between convertible existential-variables-free terms.
+
+## Simplification of atomic unification problems
+
+An atomic unification problem can be simplified as follows:
+- Variables of the context of the existential variables that are bound to terms mentioning variables of the context of the equations in non-erasable position in only one of the two instances can be restricted since there would be no way to make these terms match a subterm of the instance of the other existential variables; otherwise said, both existential variables can be restricted on the (well-typed closure of the) subsets of their context which intersect.
```

Comment:
What does "non-erasable position" mean?  What does "restricted" mean?

Reply:
If I'm understanding correctly, this again seems incorrect for the same reason as above.  We might have consider `x : nat |- ?e : bool` and `x : nat |- ?e' : bool` and the triple of unification problems `x, y : nat |- ?e@{x:=S x} == ?e'@{x:=S y}` and `x : nat |- ?e@{x:=x} == match x with O => true | S _ => false end` and `x : nat |- ?e@{x:=x} == match x with O => true | S _ => false end`. We must instantiate `?e` and `e'` both with `match x with O => true | S _ => false end`.
```coq
Check let e := (fun x => ?[e]) in
      let e' := (fun x => ?[e']) in
      fun x y => eq_refl
                 : (e (S x), e, e') =
                   (e' (S y), (fun x => match x with O => true | S _ => false end),
                    (fun x => match x with O => true | S _ => false end)).
(* Error:
In environment
e := fun x : nat => match x with
                    | 0 => true
                    | S _ => false
                    end : nat -> bool
e' := fun _ : nat => false : nat -> bool
x : nat
y : nat
The term "eq_refl" has type "(e (S x), e, e') = (e (S x), e, e')" while it is expected to have type
 "(e (S x), e, e') =
  (e' (S y), fun x : nat => match x with
                            | 0 => true
                            | S _ => false
                            end, fun x : nat => match x with
                                                | 0 => true
                                                | S _ => false
                                                end)" (cannot unify "e'" and
"fun x : nat => match x with
                | 0 => true
                | S _ => false
                end").
*)
```